File: apps/server/src/chat/claude-fork.ts
import { existsSync } from 'node:fs';
import { mkdir, readdir, readFile, rm } from 'node:fs/promises';
import { join } from 'node:path';
import { isNotFound, writeAtomic } from '../fs.ts';
import { claudeProjectSlug } from './claude-transcript.ts';
import { ChatError } from './errors.ts';

/* The Claude Code whose transcript this cut was checked against; the file is undocumented, so a change there is a refusal here. */
export const CLAUDE_TRANSCRIPT_CHECKED = '2.1.273';

/*
 * Where a fork is cut: after the line of the turn's last answer when the CLI named it, after the
 * first `turns` prompts when it did not (an older turn), or nowhere for the last turn of the chat.
 */
export type TranscriptCutPoint = { lastUuid: string } | { turns: number } | 'whole';

type Line = Record<string, unknown>;

const isRecord = (value: unknown): value is Line => typeof value === 'object' && value !== null && !Array.isArray(value);

const formatRefusal = (detail: string): ChatError =>
    new ChatError(
        'transcript-format',
        `Claude Code's conversation file has a shape this machine does not know (${detail}; checked against Claude Code ${CLAUDE_TRANSCRIPT_CHECKED}), so it cannot be forked`
    );

/* A prompt of the person or a wake, as opposed to a tool result, an injected skill or the summary a compaction leaves. */
const isSpokenPrompt = (line: Line): boolean => {
    if (line.type !== 'user' || line.isSidechain === true || line.isMeta === true || line.isCompactSummary === true) {
        return false;
    }
    const message = isRecord(line.message) ? line.message : null;
    const content = message?.content;
    if (typeof content === 'string') {
        return content !== '';
    }
    return (
        Array.isArray(content) &&
        content.some((block) => isRecord(block) && block.type === 'text') &&
        !content.some((block) => isRecord(block) && block.type === 'tool_result')
    );
};

/*
 * Claude Code 2.1.273 resumes a copied transcript cut after the chosen answer. Keep that turn's
 * trailing metadata, drop the next prompt's queued lines, and rewrite every retained session id.
 */
export const cutTranscript = (text: string, at: TranscriptCutPoint, sessionId: string, newSessionId: string): string => {
    const lines: Line[] = [];
    for (const raw of text.split('\n')) {
        if (raw.trim() === '') {
            continue;
        }
        let parsed: unknown;
        try {
            parsed = JSON.parse(raw);
        } catch {
            throw formatRefusal('a line is not JSON');
        }
        if (!isRecord(parsed) || typeof parsed.type !== 'string') {
            throw formatRefusal('a line has no type');
        }
        lines.push(parsed);
    }
    const messages = lines.filter((line) => line.type === 'user' || line.type === 'assistant');
    if (messages.length === 0 || messages.some((line) => typeof line.uuid !== 'string' || line.sessionId !== sessionId || !isRecord(line.message))) {
        throw formatRefusal('its messages lack a uuid, a message or this session id');
    }

    let end = lines.length;
    if (at !== 'whole') {
        let start: number;
        if ('lastUuid' in at) {
            start = lines.findIndex((line) => line.uuid === at.lastUuid);
            if (start < 0) {
                throw formatRefusal(`no line carries the uuid ${at.lastUuid} the turn ended on`);
            }
        } else {
            let seen = 0;
            start = lines.findIndex((line) => isSpokenPrompt(line) && ++seen === at.turns);
            if (start < 0) {
                throw new ChatError(
                    'turn-not-found',
                    `Claude Code's conversation file has fewer than ${at.turns} prompts, so the turn could not be found in it`
                );
            }
        }
        const next = lines.findIndex((line, index) => index > start && isSpokenPrompt(line));
        end = next < 0 ? lines.length : next;
    }
    const lastMessage = lines.slice(0, end).findLastIndex((line) => typeof line.uuid === 'string');
    return lines
        .slice(0, end)
        .filter((line, index) => !(index > lastMessage && line.type === 'queue-operation'))
        .map((line) => `${JSON.stringify('sessionId' in line ? { ...line, sessionId: newSessionId } : line)}\n`)
        .join('');
};

/* The transcript of a session: in the folder of the chat's directory, else wherever the session id alone finds it. */
const findTranscript = async (projectsDir: string, cwd: string, sessionId: string): Promise<string | null> => {
    const own = join(projectsDir, claudeProjectSlug(cwd), `${sessionId}.jsonl`);
    if (existsSync(own)) {
        return own;
    }
    let dirs: string[];
    try {
        dirs = await readdir(projectsDir);
    } catch (e) {
        if (isNotFound(e)) {
            return null;
        }
        throw e;
    }
    return dirs.map((dir) => join(projectsDir, dir, `${sessionId}.jsonl`)).find((candidate) => existsSync(candidate)) ?? null;
};

export interface ClaudeForkInput {
    projectsDir: string;
    cwd: string;
    sessionId: string;
    at: TranscriptCutPoint;
    /* The directory the fork works in, whose folder under the projects is where `--resume` looks. */
    forkCwd: string;
    newSessionId: string;
}

/* Writes the fork's transcript and answers how to take it back when a later step of the fork is refused. */
export const forkClaudeTranscript = async (input: ClaudeForkInput): Promise<{ path: string; undo(): Promise<void> }> => {
    if (input.projectsDir === '' || /[/\\]|\.\./.test(input.sessionId)) {
        throw new ChatError('transcript-missing', 'Claude Code keeps no conversations on this machine');
    }
    const source = await findTranscript(input.projectsDir, input.cwd, input.sessionId);
    if (source === null) {
        throw new ChatError(
            'transcript-missing',
            `Claude Code's conversation file ${join(input.projectsDir, claudeProjectSlug(input.cwd), `${input.sessionId}.jsonl`)} is not there`
        );
    }
    const copy = cutTranscript(await readFile(source, 'utf8'), input.at, input.sessionId, input.newSessionId);
    const dir = join(input.projectsDir, claudeProjectSlug(input.forkCwd));
    const path = join(dir, `${input.newSessionId}.jsonl`);
    await mkdir(dir, { recursive: true, mode: 0o700 });
    await writeAtomic(path, copy);
    return { path, undo: () => rm(path, { force: true }) };
};

File: crates/ruimte_runtime/src/chat/claude_fork.rs
use std::path::{Path, PathBuf};

use serde_json::{Map, Value};
use tokio::{fs, io::AsyncWriteExt};
use uuid::Uuid;

use crate::rpc::RpcError;

// Claude Code's transcript is undocumented; this shape is pinned to 2.1.273.
const CHECKED_VERSION: &str = "2.1.273";

pub enum CutPoint<'a> {
    LastUuid(&'a str),
    Turns(usize),
    Whole,
}

pub async fn fork_transcript(
    projects_dir: &Path,
    cwd: &str,
    session_id: &str,
    at: CutPoint<'_>,
    fork_cwd: &str,
    new_session_id: &str,
) -> Result<PathBuf, RpcError> {
    if session_id.is_empty()
        || session_id.contains('/')
        || session_id.contains('\\')
        || session_id.contains("..")
    {
        return Err(RpcError::new(
            "transcript-missing",
            "Claude Code keeps no conversation for this chat",
        ));
    }
    let source = find_transcript(projects_dir, cwd, session_id)
        .await?
        .ok_or_else(|| {
            RpcError::new(
                "transcript-missing",
                format!("Claude Code's conversation file for {session_id} is not there"),
            )
        })?;
    let text = fs::read_to_string(&source)
        .await
        .map_err(|error| RpcError::new("fork-failed", error.to_string()))?;
    let copy = cut_transcript(&text, at, session_id, new_session_id)?;
    let directory = projects_dir.join(project_slug(fork_cwd));
    fs::create_dir_all(&directory)
        .await
        .map_err(|error| RpcError::new("fork-failed", error.to_string()))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&directory, std::fs::Permissions::from_mode(0o700))
            .await
            .map_err(|error| RpcError::new("fork-failed", error.to_string()))?;
    }
    let path = directory.join(format!("{new_session_id}.jsonl"));
    let temporary = directory.join(format!(".{new_session_id}.{}.tmp", Uuid::new_v4()));
    let mut options = fs::OpenOptions::new();
    options.create_new(true).write(true);
    #[cfg(unix)]
    options.mode(0o600);
    let mut file = options
        .open(&temporary)
        .await
        .map_err(|error| RpcError::new("fork-failed", error.to_string()))?;
    if let Err(error) = file.write_all(copy.as_bytes()).await {
        let _ = fs::remove_file(&temporary).await;
        return Err(RpcError::new("fork-failed", error.to_string()));
    }
    if let Err(error) = file.sync_all().await {
        let _ = fs::remove_file(&temporary).await;
        return Err(RpcError::new("fork-failed", error.to_string()));
    }
    drop(file);
    if let Err(error) = fs::rename(&temporary, &path).await {
        let _ = fs::remove_file(&temporary).await;
        return Err(RpcError::new("fork-failed", error.to_string()));
    }
    Ok(path)
}

pub fn cut_transcript(
    text: &str,
    at: CutPoint<'_>,
    session_id: &str,
    new_session_id: &str,
) -> Result<String, RpcError> {
    let mut lines = Vec::<Map<String, Value>>::new();
    for raw in text.lines().filter(|line| !line.trim().is_empty()) {
        let line = serde_json::from_str::<Value>(raw)
            .ok()
            .and_then(|value| value.as_object().cloned())
            .filter(|line| line.get("type").is_some_and(Value::is_string))
            .ok_or_else(|| format_refusal("a line is not a typed JSON object"))?;
        lines.push(line);
    }
    let messages = lines.iter().filter(|line| {
        matches!(
            line.get("type").and_then(Value::as_str),
            Some("user" | "assistant")
        )
    });
    if messages.clone().next().is_none()
        || messages.clone().any(|line| {
            !line.get("uuid").is_some_and(Value::is_string)
                || line.get("sessionId").and_then(Value::as_str) != Some(session_id)
                || !line.get("message").is_some_and(Value::is_object)
        })
    {
        return Err(format_refusal(
            "its messages lack a uuid, a message or this session id",
        ));
    }
    let end = match at {
        CutPoint::Whole => lines.len(),
        CutPoint::LastUuid(uuid) => {
            let start = lines
                .iter()
                .position(|line| line.get("uuid").and_then(Value::as_str) == Some(uuid))
                .ok_or_else(|| {
                    format_refusal(&format!(
                        "no line carries the uuid {uuid} the turn ended on"
                    ))
                })?;
            next_prompt(&lines, start).unwrap_or(lines.len())
        }
        CutPoint::Turns(turns) => {
            let mut seen = 0;
            let start = lines
                .iter()
                .position(|line| {
                    if spoken_prompt(line) {
                        seen += 1;
                    }
                    seen == turns && spoken_prompt(line)
                })
                .ok_or_else(|| {
                    RpcError::new(
                        "turn-not-found",
                        format!(
                            "Claude Code keeps fewer than {turns} prompts, so the turn could not be found"
                        ),
                    )
                })?;
            next_prompt(&lines, start).unwrap_or(lines.len())
        }
    };
    let last_message = lines[..end]
        .iter()
        .rposition(|line| line.get("uuid").is_some_and(Value::is_string));
    let mut output = String::new();
    for (index, mut line) in lines.into_iter().take(end).enumerate() {
        if last_message.is_some_and(|last| index > last)
            && line.get("type").and_then(Value::as_str) == Some("queue-operation")
        {
            continue;
        }
        if line.contains_key("sessionId") {
            line.insert("sessionId".into(), Value::String(new_session_id.into()));
        }
        output.push_str(&serde_json::to_string(&line).expect("JSON object serialization"));
        output.push('\n');
    }
    Ok(output)
}

async fn find_transcript(
    projects_dir: &Path,
    cwd: &str,
    session_id: &str,
) -> Result<Option<PathBuf>, RpcError> {
    let own = projects_dir
        .join(project_slug(cwd))
        .join(format!("{session_id}.jsonl"));
    if fs::try_exists(&own).await.unwrap_or(false) {
        return Ok(Some(own));
    }
    let mut entries = match fs::read_dir(projects_dir).await {
        Ok(entries) => entries,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(error) => return Err(RpcError::new("fork-failed", error.to_string())),
    };
    while let Some(entry) = entries
        .next_entry()
        .await
        .map_err(|error| RpcError::new("fork-failed", error.to_string()))?
    {
        let candidate = entry.path().join(format!("{session_id}.jsonl"));
        if fs::try_exists(&candidate).await.unwrap_or(false) {
            return Ok(Some(candidate));
        }
    }
    Ok(None)
}

fn project_slug(cwd: &str) -> String {
    cwd.chars()
        .map(|character| {
            if character.is_ascii_alphanumeric() {
                character
            } else {
                '-'
            }
        })
        .collect()
}

fn next_prompt(lines: &[Map<String, Value>], start: usize) -> Option<usize> {
    lines
        .iter()
        .enumerate()
        .skip(start + 1)
        .find_map(|(index, line)| spoken_prompt(line).then_some(index))
}

fn spoken_prompt(line: &Map<String, Value>) -> bool {
    if line.get("type").and_then(Value::as_str) != Some("user")
        || line.get("isSidechain").and_then(Value::as_bool) == Some(true)
        || line.get("isMeta").and_then(Value::as_bool) == Some(true)
        || line.get("isCompactSummary").and_then(Value::as_bool) == Some(true)
    {
        return false;
    }
    let Some(content) = line
        .get("message")
        .and_then(|message| message.get("content"))
    else {
        return false;
    };
    if content.as_str().is_some_and(|text| !text.is_empty()) {
        return true;
    }
    let Some(blocks) = content.as_array() else {
        return false;
    };
    blocks
        .iter()
        .any(|block| block.get("type").and_then(Value::as_str) == Some("text"))
        && !blocks
            .iter()
            .any(|block| block.get("type").and_then(Value::as_str) == Some("tool_result"))
}

fn format_refusal(detail: &str) -> RpcError {
    RpcError::new(
        "transcript-format",
        format!(
            "Claude Code's conversation file has a shape this machine does not know ({detail}; checked against Claude Code {CHECKED_VERSION}), so it cannot be forked"
        ),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cuts_after_named_turn_and_rewrites_session() {
        let line = |value: Value| format!("{}\n", serde_json::to_string(&value).unwrap());
        let transcript = [
            line(serde_json::json!({"type":"user","uuid":"u1","sessionId":"old","message":{"content":"one"}})),
            line(serde_json::json!({"type":"assistant","uuid":"a1","sessionId":"old","message":{"content":[]}})),
            line(serde_json::json!({"type":"queue-operation","operation":"keep"})),
            line(serde_json::json!({"type":"user","uuid":"u2","sessionId":"old","message":{"content":"two"}})),
            line(serde_json::json!({"type":"assistant","uuid":"a2","sessionId":"old","message":{"content":[]}})),
        ]
        .concat();
        let cut = cut_transcript(&transcript, CutPoint::LastUuid("a1"), "old", "new").unwrap();
        let lines = cut.lines().collect::<Vec<_>>();
        assert_eq!(lines.len(), 2);
        assert!(lines[0].contains("\"sessionId\":\"new\""));
        assert!(!cut.contains("queue-operation"));
        assert!(!cut.contains("a2"));
    }

    #[test]
    fn drops_trailing_queue_operations() {
        let transcript = [
            serde_json::json!({"type":"user","uuid":"u1","sessionId":"old","message":{"content":"one"}}),
            serde_json::json!({"type":"assistant","uuid":"a1","sessionId":"old","message":{"content":[]}}),
            serde_json::json!({"type":"queue-operation","operation":"drop"}),
        ]
        .into_iter()
        .map(|value| format!("{}\n", serde_json::to_string(&value).unwrap()))
        .collect::<String>();
        let cut = cut_transcript(&transcript, CutPoint::Whole, "old", "new").unwrap();
        assert!(!cut.contains("queue-operation"));
    }
}
